### ISSUE TITLE
[Bug] Jumping categories in filter bottom sheet

### ISSUE THREAD

**Issue by patrickunterwegs** [TechbeeAT/jtxBoard#827]
@patrickunterwegs 
Slight problem with showing categories by most used or recently used - on my system (v2.05.03) it is now sometimes taking about 30 secs after opening the filter dialogue for the categories section to stop jumping around. It is obviously doing this as it trawls through all of the items updating the recent/most used counts and changing the display as it progresses. It doesn't always do this, but about one time in three - possibly when there are new items since the filter dialogue was last opened, or when the app has been closed and re-opened.

This makes the filter dialogue category selection very unstable as the order continually flashes around so you can select. 30secs is too long to wait.

With 400+ items in the list and 30+ categories in use this is taking over 30secs which is not really usable - could it do it in the background and simply change the order next time the filter dialogue is opened rather than messing on the live screen?

(or switch to a static user selected 5 most wanted categories which would then never change - actually I would find this most helpful, to be able to choose 5 categories which would stick to the top of the list and the rest (more...) alphabetical. Then you'd never need to trawl through re-counting.)

_Originally posted by @rogercreagh in https://github.com/TechbeeAT/jtxBoard/issues/802#issuecomment-1693120512_

**Comment by patrickunterwegs** [comment]
Hey @rogercreagh, 
hm, okay, I haven't seen the problem, but I can imagine what causes it. 

So here's a short technical explanation as you might be interested: 
In Android there's a library called "room" that takes care of databases and transactions. The common pattern is to load data in a "LiveData" object, that is dynamically loaded - initially and whenever something changes. This LiveData object can be observed in the UI and refreshed the UI, whenever the content of the LiveData object changes. This makes it very convenient to use and it's also the most modern practice to handle fetching data from a database.

Another concept in Android/Jetpack compose is to use LazyLists. LazyLists are lists that can contain many entries. The UI element takes care that only elements are rendered, that are actually visible on the screen making it fast to show and scoll. In the best case LiveData and LazyLists can be combined. 

In the case of categories it uses LiveData, but then uses a FlowRow (where elements just take as much space as possible and the next one is next to it or in the next row). But this FlowRow doesn't support this "Lazy" concept yet. This might cause the list to always be fully rendered (which might also cause the flickering). 

Basically are you sure that it isn't the sync running in the background that causes the flickering? Data fetching should never take 30 seconds...

I have 2 approaches to get it resolved: 
- mark the category fetching as a transaction in the database query
- use another kind of list (for example a horizontally scrollable lazy row)

I will try the first one in the next version, if it doesn't help, I'll check the second.

**Comment by rogercreagh** [comment]
Hmm, I don't think a sync was running when it happens, I'll check carefully next time - I suppose if a sync effectively momentarily deletes and replaces items with a fresh copy then it might cause the LiveData to update as each item gets replaced?

The effect looks like exactly what I would expect if some background process was trawling through each entry in turn and updating the used count and date for any categories assigned to each item in turn and every time the the count/date is updated then the 5 categories displayed change (which would be LiveData updating I guess) - its not just a flickering, the actual categories displayed change and reorder.

I'm not at all clear how sync through Davx5 works. I have two and sometimes three devices using the collection so if a different device has run sync then ctag will be updated and sync on the device I am looking at may re-fetch all the items if it is using simple CalDAV sync or perhaps just the changed ones if it is using WebDAV sync - not sure what DavX5 does.

When the bug is present I do get the impression that it is re-indexing  the entire data set, not just ones that have changed in a sync.

**Comment by patrickunterwegs** [comment]
Currently the sync would update the main entry, but all attributes that can occur multiple times (like categories) are deleted and inserted again.
That would also explain the behaviour you describe in your second paragraph. 
Once 2.5.4 is available for you, please just check if the change helped. If not (actually I suppose that it will work just a little better but the problem persists), I will think of another solution to load the categories only once when opening the screen (ie. also doing it without LiveData)

**Comment by rogercreagh** [comment]
OK I've been using 2.5.4 for a few days on two devices and not seen a recurrence - I think this can be closed.

**Comment by patrickunterwegs** [comment]
Great, thank you for the update =)
Anyway, if it occurs again, we can still reopen it. 👍

**Comment by patrickunterwegs** [comment]
Hi @rogercreagh , just wanted to check once more, did the jumping categories issue occur again?
I prepared another solution that I don't like, so I'd be happy if the current solution really fixed the problem 😉

**Comment by rogercreagh** [comment]
HI @patrickunterwegs I haven't seen it with 2.5.5 so I think whatever you did has done it. I'm about to disappear offline for 10 days, but I'll still be using jtxB daily but won't be able to post responses here.

Separately I'm having a nightmare trying to set up GitHub's 2FA which I have to do by early Oct or get locked out. As a fallback I'll be moving all my stuff off GitHub. 2FA is a frigging nightmare and serves no purpose for the user. Goodness knows what GitHub get out of it (apart from pissed off ex-Users!)

**Comment by patrickunterwegs** [comment]
Alright, thank you for the update! I hope it'll be gone for good 😊
